operator/v1alpha1/types_kueue.go

```diff
@@ -0,0 +1,277 @@
+	// When admission is set to Sequential, only pods from the currently processing workload will be admitted.
+	// Once all pods from the current workload are admitted, and ready, Kueue will process the next workload.
+	// Sequential processing may slow down admission when the cluster has sufficient capacity for multiple workloads,
+	// but provides a higher guarantee of workloads scheduling all pods together successfully.
+	// When set to Parallel, pods from any workload will be admitted at any time.
+	// This may lead to a deadlock where workloads are in contention for cluster capacity and
+	// pods from another workload having successfully scheduled prevent pods from the current workload scheduling.
+	// +kubebuilder:validation:XValidation:rule="self.policy==ByWorkload",message="byWorkload is only valid if policy equals ByWorkload"
+	// +optional
+	ByWorkload KueueGangSchedulingAdmissionOptions `json:"byWorkload"`
+}
+
+// +kubebuilder:validation:Enum=QueueNameRequired;QueueNameOptional
+type QueueLabelNamePolicy string
+
+const (
+	QueueLabelNamePolicyRequired QueueLabelNamePolicy = "QueueNameRequired"
+	QueueLabelNamePolicyOptional QueueLabelNamePolicy = "QueueNameOptional"
+)
+
+type QueueLabelPolicy struct {
+	// queueLabelPolicy controls whether or not Kueue reconciles
+	// jobs that don't set the label kueue.x-k8s.io/queue-name.
+	// The allowed values are QueueNameRequired and QueueNameOptional.
+	// If set to QueueNameRequired, then those jobs will be suspended and never started unless
+	// they are assigned a queue and eventually admitted. This also applies to
+	// jobs created before starting the kueue controller.
+	// Defaults to QueueNameRequired; therefore, those jobs are not managed and if they are created
+	// unsuspended, they will start immediately.
+	// +optional
+	QueueLabelPolicy QueueLabelNamePolicy `json:"queueLabelPolicy"`
+}
+
+// +kubebuilder:validation:Enum=Classical;FairSharing
+type PreemptionStrategy string
+
+const (
+	PreemeptionStrategyClassical   PreemptionStrategy = "Classical"
+	PreemeptionStrategyFairsharing PreemptionStrategy = "FairSharing"
+)
+
+type Preemption struct {
+	// preemptionStrategy are the types of preemption kueue allows.
+	// Kueue has two types of preemption: classical and fair sharing.
+	// Classical means that an incoming workload, which does
+	// not fit within the unusued quota, is eligible to issue preemptions
+	// when the requests of the workload are below the
+	// resource flavor's nominal quota or borrowWithinCohort is enabled
+	// on the Cluster Queue.
+	// FairSharing is a more heavy weight algorithm.
+	// ClusterQueues with pending Workloads can preempt other Workloads
+	// in their cohort until the preempting ClusterQueue
+	// obtains an equal or weighted share of the borrowable resources.
+	// The borrowable resources are the unused nominal quota
+	// of all the ClusterQueues in the cohort.
+	// +optional
+	PreemptionStrategy PreemptionStrategy `json:"preemptionStrategy"`
```

Comment:
Looks like the upstream configuration allows configuring multiple preemption strategies that run in a chain. Are we not planning to allow that and forcing the choice of one or the other?

Reply:
https://github.com/openshift/api/pull/2250#discussion_r2015569569

Reply:
TLDR it is really hard to explain those policies and the customer ask right now is to enable the kueue defaults for FairSharing. Not to provide all possible items.

Reply:
I was thinking I can add a new enum and set the values I need without having to explain the kueue APIs as is.

Reply:
>I was thinking I can add a new enum and set the values I need without having to explain the kueue APIs as is.

Am I understanding this correctly as maintaining a set of enums for every permutation of the preemption strategies?

I suspect that is going to be troublesome and would very much so advise against that.

I think the appropriate way would be to introduce an optional struct in the future that exposes the ability to configure the ordered list of fair sharing algorithms

Reply:
I haven't thought that hard about this but I turning this into a discriminated union would be the right approach.

Reply:
We don't have a request for this though. And if we do we can always define a enum enum and set the values in our operator. Otherwise we have to expose Kueue APIs which is the exact thing we want to avoid.

Reply:
We have 1 ask for enabling FairSharing from IBM and they are happy with the defaults. If someone really wants to tune it and figure out a better strategy I think we should push them to engage with upstream. And we can always default to a setting in our controller.

I really want to avoid having to explain and document a very complicated and quickly changing part of the API.

Reply:
I don't think we are saying you have to support it now. We are looking to get an idea of how, if a request for this comes in, you plan to enable that functionality in a non-breaking way.

Reply:
>If someone really wants to tune it and figure out a better strategy I think we should push them to engage with upstream

Am I correct in understanding this as essentially boiling down to "if you don't like our opinions, use the upstream version"?

Reply:
Not quite. So upstream has this to say about this Fair sharing policies.


> Different preemptionStrategies can lead to less or more preemptions under specific scenarios. These are the factors you should consider when configuring preemptionStrategies:
> 
> - Tolerance to disruptions, in particular when single Workloads use a significant amount of the borrowable resources.
> - Speed of convergence, in other words, how important is it to reach a steady fair state as soon as possible.
> - Overall utilization, because certain strategies might reduce the utilization of the cluster in the pursue of fairness.

Reply:
The other complication is https://github.com/kubernetes-sigs/kueue/pull/4252/files.

Looking at this, it actually seems that maybe our Preemption Policy may not be the best name as this would be fair sharing without any prememption.